File: software/examples/ruby/example_callback.rb
#!/usr/bin/env ruby
# -*- ruby encoding: utf-8 -*-

# FIXME: This example is incomplete

require 'tinkerforge/ip_connection'
require 'tinkerforge/bricklet_dc_v2'

include Tinkerforge

HOST = 'localhost'
PORT = 4223
UID = 'XYZ' # Change XYZ to the UID of your DC Bricklet 2.0

ipcon = IPConnection.new # Create IP connection
dc = BrickletDCV2.new UID, ipcon # Create device object

ipcon.connect HOST, PORT # Connect to brickd
# Don't use device before ipcon is connected

# The acceleration has to be smaller or equal to the maximum
# acceleration of the DC motor, otherwise the velocity reached
# callback will be called too early
dc.set_motion 4096, \
              16384 # Slow acceleration (12.5 %/s), fast decceleration (50 %/s) for stopping
dc.set_velocity 32767 # Full speed forward (100 %)

# Use velocity reached callback to swing back and forth
# between full speed forward and full speed backward
dc.register_callback(BrickletDCV2::CALLBACK_VELOCITY_REACHED) do |velocity|
  if velocity == 32767
    puts 'Velocity: Full speed forward, now turning backward'
    dc.set_velocity(-32767)
  elsif velocity == -32767
    puts 'Velocity: Full speed backward, now turning forward'
    dc.set_velocity 32767
  else
    puts 'Error' # Can only happen if another program sets velocity
  end
end

# Enable motor power
dc.set_enabled true

puts 'Press key to exit'
$stdin.gets

dc.set_velocity 0 # Stop motor before disabling motor power
sleep 2 # Wait for motor to actually stop: velocity (100 %) / decceleration (50 %/s) = 2 s
dc.set_enabled false # Disable motor power

ipcon.disconnect

File: software/examples/php/ExampleCallback.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletDCV2.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletDCV2;

const HOST = 'localhost';
const PORT = 4223;
const UID = 'XYZ'; // Change XYZ to the UID of your DC Bricklet 2.0

// Use velocity reached callback to swing back and forth
// between full speed forward and full speed backward
function cb_velocityReached($velocity, $user_data)
{
    $dc = $user_data;

    if ($velocity == 32767) {
        echo "Velocity: Full speed forward, now turning backward\n";
        $dc->setVelocity(-32767);
    } elseif ($velocity == -32767) {
        echo "Velocity: Full speed backward, now turning forward\n";
        $dc->setVelocity(32767);
    } else {
        echo "Error\n"; // Can only happen if another program sets velocity
    }
}

$ipcon = new IPConnection(); // Create IP connection
$dc = new BrickletDCV2(UID, $ipcon); // Create device object

$ipcon->connect(HOST, PORT); // Connect to brickd
// Don't use device before ipcon is connected

// The acceleration has to be smaller or equal to the maximum
// acceleration of the DC motor, otherwise the velocity reached
// callback will be called too early
$dc->setMotion(4096,
               16384); // Slow acceleration (12.5 %/s), fast decceleration (50 %/s) for stopping
$dc->setVelocity(32767); // Full speed forward (100 %)

// Register velocity reached callback to function cb_velocityReached
$dc->registerCallback(BrickletDCV2::CALLBACK_VELOCITY_REACHED, 'cb_velocityReached', $dc);

// Enable motor power
$dc->setEnabled(TRUE);

echo "Press ctrl+c to exit\n";
$ipcon->dispatchCallbacks(-1); // Dispatch callbacks forever

?>
